feat: add support for partial content

juanscolari opened the pull request:
## Description

Support partial fetch to download content when sending the `Range` header

Ignore the ranges when they are invalid, or the content is encoded

Closes [#896](https://github.com/decentraland/catalyst/issues/896)

Docs: https://github.com/decentraland/catalyst-api-specs/pull/23

## Changes

- Status Code
  - Set Status Code `206` only when requesting partial content. (`200` is still returned for the whole file)
- Request Headers
  - Parse `Range` header to support partial content
- Response Headers
  - Set `Accept-Ranges: bytes` header to indicate the client that there is support por partial content
  - Set `Content-Length` header accordingly. Por partial content, it's the size of the chunk. Por full files it's the size of the whole file
  - Set `Content-Range` header for partial content, describing the chunk and the total file size
- Response
  - Return the partial response when a `Range` header is provided

## Out of Scope

- Lack of support for [416](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/416) Status Code errors. The current behavior is to return the whole file, as if no `Range` was provided.

## Supporting Documentation

- [Range Requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests)
- [Partial Content](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/206)
- [Range](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Range)
- [Accept-Ranges](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Ranges)
- [Content-Range](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Range)
- [w3 Org Spec](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.35.1)

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change). If there's an API change, link the pull request in the [API spec repository](https://github.com/decentraland/catalyst-api-specs) and the accepted [DAO governance poll](https://governance.decentraland.org/)
- [x] I have read the [**CONTRIBUTING**](https://github.com/decentraland/catalyst/blob/main/docs/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

agusaldasoro commented on content/src/controller/Controller.ts:
```diff
@@ -227,6 +232,28 @@ export class Controller {
+    // Set default values when they are not provided
+    const ranges = rangeHeader.substring(6, rangeHeader.length).split('-')
+    const start = ranges[0] ? Number(ranges[0]) : 0
+    const end = ranges[1] ? Number(ranges[1]) : Number.MAX_SAFE_INTEGER
```
what happens when the end is bigger than the file size?

agusaldasoro commented on content/src/storage/FileSystemContentStorage.ts:
```diff
@@ -53,19 +54,31 @@ export class FileSystemContentStorage implements ContentStorage {
       const stat = await fs.promises.stat(filePath)
+      if (range && stat.size) {
+        const safeRange = {
+          start: Math.min(range.start, stat.size - 1),
```
shouldn't this be `Math.max(range.start, 0)` ?

juanscolari replied:
I made it that way initially to prevent having a `range.start` bigger than the file size. I've now refactored the code to make it clearer, setting some default values when the ranges are not provided and validating them later

guidota commented on content/src/storage/FileSystemContentStorage.ts:
```diff
@@ -53,19 +54,42 @@ export class FileSystemContentStorage implements ContentStorage {
+      if (range && stat.size) {
+        // Set default values for ranges when they are not defined
+        const rangeStart = range.start === undefined ? 0 : range.start
+        const rangeEnd = range.end === undefined ? stat.size - 1 : range.end
```
```suggestion
        const rangeEnd = range.end ?? stat.size - 1
```

guidota replied:
nit:

guidota commented on content/src/controller/Controller.ts:
```diff
@@ -669,12 +691,21 @@ async function setContentFileHeaders(content: RawContent, hashId: string, res: e
   }
 
-  if (content.size) {
+  if (content.size && content.range?.start !== undefined && content.range?.end !== undefined) {
```
is it possible to `start` or `end` be undefined but still a valid range?

guidota replied:
ok, it is not possible but maybe is a good idea to have the optional range but required `start`/`end` when you create the `ContentItem`

juanscolari replied:
Thanks for pointing out this. I was able to come up with a more elegant solution here: https://github.com/decentraland/catalyst/pull/915/commits/67397820ff56ce06a1a7db63887c6a532a9d703f

The point was that the `range` that is used to create the headers is one that is calculated when reading the file. So, we have both start and end properties or none of them.

guidota reviewed (commented):
nice 🙌

menduz commented:
Quick question, I'm AFK so I couldn't go through the code. 

will it work the same for compressed assets like the snapshot files?

juanscolari commented:
@menduz 
> will it work the same for compressed assets like the snapshot files?

Good catch. I've removed range support for files with encoding as it didn't work. I've also removed the header indicating support for range request for those files.

The current behavior  for files with encoding is to ignore the range headers and return the whole file with a `200` status code.

https://github.com/decentraland/catalyst/pull/915/commits/40ed52c6131897350e8f753be091ccc51d0fb631
